File: Scala Actors Vs Java Threads - Gossip and Push Sum Algorithms/Project2.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.LinkedList;
import java.util.List;
import java.util.Queue;
import java.util.Random;
import java.util.Set;




public class project2 implements Runnable{
	private int id;
	private double s;
	private double w;

		project2(int id)
		{
			this.id=id;
			this.s = id;
			this.w = 1;
	    }	
		static ArrayList<Long> input = new ArrayList<Long>();
		static ArrayList<Long> output = new ArrayList<Long>();
		static int numThreads;
		static double numNodes,numNodesRoot;
		static int maxRumours,rumourCount;
		static int selfRumourCount;
		static boolean exit1 = false;
		static long N;
		static int nodeId;
		static String topology,algorithm;
		static int first;
		static List<Queue<Integer>> queues = new ArrayList<Queue<Integer>>();
		static List<Queue<double[]>> queuespushsum = new ArrayList<Queue<double[]>>();
		static Set<Object> sendList=new HashSet<Object>();
		static Set<Object> deleteList=new HashSet<Object>();
		static Set<Integer> not_received_any=new HashSet<Integer>();		
	
	public static void main(String[] args) {

		if(args.length==3)
		{
			numNodes=Integer.parseInt(args[0]);
			topology=args[1];
			algorithm = args[2];
			maxRumours = 10;
			
			if(topology == "2d" || topology == "imp2d")
			      while(Math.sqrt(numNodes)%1!=0){
			       numNodes += 1;
			     }
			numNodesRoot = Math.sqrt(numNodes);
			numThreads = (int) numNodes;
			for(int i=0;i<numThreads;i++){
				Queue<Integer> q = new LinkedList<Integer>();
				queues.add(q);
			}
			for(int i=0;i<numThreads;i++){
				Queue<double[]> q = new LinkedList<double[]>();
				queuespushsum.add(q);
			}
			Random rand = new Random();
            first = rand.nextInt(numThreads);
            for(int i=0;i<numThreads;i++)
            	not_received_any.add(i);
            long starttime = System.currentTimeMillis();
            gossipSimulator();
            long totaltime = (System.currentTimeMillis() - starttime);
            System.out.println("Time taken: "+totaltime+"ms");
			
		    
		}
		else
			System.out.print("error :: Please provide the value of N and k as command line arguments.");
	}
		
	private static void gossipSimulator() {
		ArrayList<Thread> arr = new ArrayList<>();

		for(int i = 0;i < numThreads ; i++){
		Thread t = new Thread(new project2(i));	
		
		arr.add(t);
		}
		for(int i=0;i<numThreads;i++){
			arr.get(i).start();
		}
		for(int i=0;i<numThreads;i++){
		    try{
		    	arr.get(i).join();
		    
		    }catch(InterruptedException e){
		    	 e.printStackTrace();
		    }
			
		}
	
	}

	@Override
	public void run() {
		 ArrayList<Integer> neighbors = new ArrayList<Integer>();
		 double [] values=new double[2];
		 int rumour = 0;
		 int count = 0;
		 double sum = 0;  
		 double weight = 1;
		 if(topology.equals("full"))
		   {
		     int i = id;
		            
		            int j = 0;
		            while(j<numThreads){
		              if(j!=i){
		                neighbors.add(j);
		               		}
		              j += 1;
		            }
		   }
		if(topology.equals("2d"))
		   {
		    
		    int i = id;
		            
		            if((i - numNodesRoot) >= 0)
		              neighbors.add((int) (i-numNodesRoot));
		             if((i + numNodesRoot) < numThreads)
		             neighbors.add((int) (i+numNodesRoot));
		             if((i+1) % numNodesRoot != 0)
		               neighbors.add((int) (i));
		             if(i % numNodesRoot != 0)
		               neighbors.add((int) (i-1));  
		   }
		   
		   if(topology.equals("imp2d"))
		   {
		     
		     int i = id;
		     int randomNeighbor = -1;
		            
		            if((i - numNodesRoot) >= 0)
		              neighbors.add((int) (i-numNodesRoot));
		             if((i + numNodesRoot) < numThreads)
		             neighbors.add((int) (i+numNodesRoot));
		             if((i+1) % numNodesRoot != 0)
		               neighbors.add(i);
		             if(i % numNodesRoot != 0)
		               neighbors.add(i-1);  
		            
		            
		            do{
		            	Random rand = new Random();
		              randomNeighbor = rand.nextInt(numThreads);
		              
		                if(neighbors.contains(randomNeighbor))
		                	randomNeighbor = -1;
		            }while(randomNeighbor == -1);
		            
		            neighbors.add(randomNeighbor);
		            
		          }
		     
		   
		   if(topology.equals("line"))
		   {
		          int i = id;
		          if(i>0) neighbors.add(i-1);
		          if(i<numThreads-1) neighbors.add(i+1);
		            
		            
		     }
		   if(algorithm.equalsIgnoreCase("gossip"))
		   {
			 
			   if(id==first)
			   {
				   //Incrementing rumor
				   rumour++;
				   //add to sendlist 
				   sendList.add(Thread.currentThread());
				   // remove from not_received_any
				   not_received_any.remove(id);
				}
			   
			   // if not_received_any is empty then exit
			   while(!not_received_any.isEmpty())
			   {
				   //sending the rumour
				   if(sendList.contains(Thread.currentThread()) && rumour<10)
				   {
					    Random rand1 = new Random();
				            int receiver = rand1.nextInt(neighbors.size());
				            synchronized(queues.get(neighbors.get(receiver))){
				            	queues.get(neighbors.get(receiver)).add(id);
				             }
				            synchronized(queues.get(id)){
				            	queues.get(id).add(id);
				            }
				      
					}
				   
				   // receiving the rumour and updating the rumour count 
				   if(!queues.get(id).isEmpty())
				   {
					   int received;
					   synchronized(queues.get(id)){
					   received = queues.get(id).poll();
					   }
					  
					 //Incrementing rumor
					   if(received != id)
					   rumour++;

					  System.out.println("");
					   
					   //adding to delete list if rumour exceeds 10
					   if(rumour==10){
						   deleteList.add(Thread.currentThread()) ;
						   sendList.remove(Thread.currentThread());
					   }
					   
					   //adding if new element 
					   if(!((sendList.contains(Thread.currentThread()))&& (deleteList.contains(Thread.currentThread()))))
						{
							   sendList.add(Thread.currentThread()) ;
							   not_received_any.remove(id);
						}
					}
			   }
		   }
		   else if(algorithm.equalsIgnoreCase("pushsum")|| algorithm.equalsIgnoreCase("push-sum"))
		   {
			 
			   if(id==first)
			   {
				   //Condition initialization
				   s +=sum;
				   w+=weight;
				   s = s/2;                 
				   w = w/2;   
				   Random rand1 = new Random();
			       int receiver = rand1.nextInt(neighbors.size());
			       values[0] = s;
			       values[1] = w;
			       synchronized(queuespushsum.get(neighbors.get(receiver))){
			       queuespushsum.get(neighbors.get(receiver)).add(values);
			       }
				   rumour++;     //Incrementing rumor
				   
				  
				   sendList.add(Thread.currentThread());
			       not_received_any.remove(id);  
			   }
			   			 
			   while(!(not_received_any.isEmpty() && exit1))
			     {
				   if(!queuespushsum.get(id).isEmpty())
				       {
					   rumour++;
					   not_received_any.remove(id);
					   if(!sendList.contains(Thread.currentThread()))
						   sendList.add(Thread.currentThread());
					   double condition = Math.pow(10, -10);
					   synchronized(queuespushsum.get(id)){
					   sum = queuespushsum.get(id).peek()[0];
					   weight = queuespushsum.get(id).peek()[1];
					   queuespushsum.get(id).remove();
					   }
					   double oldRatio = s/w;   //Old ratio
					   s +=sum;
					   w +=weight;
					   s = s/2;                 
					   w = w/2;                 
					   double newRatio = s/w;   //New Ratio
					   double actor_ratio = Math.abs(oldRatio - newRatio);
					   if(actor_ratio > condition)
					   {
						   count = 0;
						   
						               
						   Random rand1 = new Random();
					       int receiver = rand1.nextInt(neighbors.size());
					       values[0] = s;
					       values[1] = w;
					       synchronized(queuespushsum.get(neighbors.get(receiver))){
					       queuespushsum.get(neighbors.get(receiver)).add(values);
					       }
						}
					   else
					   {
						   
						   if(!sendList.contains(Thread.currentThread()))
							   sendList.add(Thread.currentThread());
						   count+=1;
						   
						   if(count == 3)
						   {
							   exit1 = true; // check termination condition 
							   Random rand1 = new Random();
						       int receiver = rand1.nextInt(neighbors.size());
						       values[0] = s;
						       values[1] = w;
						       synchronized(queuespushsum.get(neighbors.get(receiver))){
						       queuespushsum.get(neighbors.get(receiver)).add(values);
						       }
						       break;
						   }
						   else
						  {
							                 
							   Random rand1 = new Random();
						       int receiver = rand1.nextInt(neighbors.size());
						       values[0] = s;
						       values[1] = w;
						       synchronized(queuespushsum.get(neighbors.get(receiver))){
						       queuespushsum.get(neighbors.get(receiver)).add(values);
						       }
						   }
					   }
				   }
				   else  if(sendList.contains(Thread.currentThread()))
					 {
						   Random rand1 = new Random();
					       int receiver = rand1.nextInt(neighbors.size());
					       values[0] = s;
					       values[1] = w;
					       synchronized(queuespushsum.get(neighbors.get(receiver))){
					       queuespushsum.get(neighbors.get(receiver)).add(values);
					       }
					 } 
				}
			}
		   else
		   {
			   System.out.println("Enter correct Algorithm.");
		   }
		
	}

}

File: Scala Actors Vs Java Threads - Gossip and Push Sum Algorithms/Project2.scala
import akka.actor._
import scala.util.Random


case class Identity(uniqueId:Int, actors:List[ActorRef], neighbors:List[Int], maxNoOfRumours:Int, superVisor:ActorRef)
case class Gossip(Id:Int)
case class Done(nodeId:Int,numNodes:Double)
case class KeepTime(b:Long)
case class pushSum(sum:Double,weight:Double)
case class PushSumDone(nodeId:Int,numNodes:Double,value:Int)

case class Nodes() extends Actor {
  
   var maxRumours: Int = 0
   var nodeId:Int = 0
   var neighbors:List[Int] = Nil
   var actors:List[ActorRef] = Nil
   var superVisor:ActorRef = null
   var selfRumourCount:Int = 0
   var count:Int = 0
   var flag :Int = _
   var numNodes:Double = 0
   var rumorCount:Int = 0
   var s:Double = 0
   var w:Double = 0
   
   def receive = {
        case Identity(nodeId:Int, actors:List[ActorRef], neighbors:List[Int], maxNoOfRumours:Int, superVisor:ActorRef) => {
  
            this.superVisor = superVisor
        	this.nodeId = nodeId
        	this.s = nodeId
        	this.w = 1
            this.neighbors = neighbors      
            this.actors = actors
            numNodes = actors.length
        	
        }

   
   case Gossip(rumourId:Int) => {
     
     if(maxRumours < 10)
       {
       if(nodeId != rumourId){
       maxRumours += 1
       if(maxRumours == 1){
       superVisor ! Done(nodeId,numNodes)
       }
       }
       
       //println("Last Rumour sent by " +rumourId+" Total No. Of Rumours received by " +nodeId+ " are " +maxRumours)
       
       var randomNode = Random.nextInt(neighbors.length)
       actors(neighbors(randomNode)) ! Gossip(nodeId)
       self ! Gossip(nodeId)
     }
    
     
     /*if(maxRumours == 10 && nodeId != rumourId)
     {
       //println(" actor "+nodeId+" Terminating")
       superVisor ! Done(nodeId,numNodes)
       context.stop(self)
     }
     
     else if(selfRumourCount == maxRumours)
     {
       count += 1
       if(count == 20)
       {
          //println(" actor "+nodeId+" Terminating")
          superVisor ! Done(nodeId,numNodes)
          context.stop(self)
       }
     }*/
     }
   
   case pushSum(sum:Double, weight:Double) => 
          {
          var condition = math.pow(10, -10)
          rumorCount+=1
          
          var oldRatio:Double = s/w;
          s+=sum
	      w+=weight
	      s=s/2
	      w=w/2
	      var newRatio:Double = s/w
	      var actor_ratio:Double = Math.abs((oldRatio-newRatio))
	      
	      if(rumorCount==1){
	        flag = 0
	        superVisor ! PushSumDone(nodeId,numNodes,flag)
	       }
	       

	      if(rumorCount==1 || actor_ratio > condition) 
          {
        	  count = 0;
        	  
	            var randomPlayer = Random.nextInt(neighbors.length);
	            actors(neighbors(randomPlayer)) ! pushSum(s,w)
	           
	      }
	      
          else
          {
           
            count+=1;
            //println("Count = "+count)
            if(count == 3) 
            {
              //println("Final Condition --- For Actor  "+nodeId+" \tRumor Count  "+rumorCount+" \ts/w: "+(s/w));
              flag = 1
              
              var randomPlayer = Random.nextInt(neighbors.length);
              actors(neighbors(randomPlayer)) ! pushSum(s,w)
              superVisor ! PushSumDone(nodeId,numNodes,flag)
	          self ! PoisonPill
	        }
            
            else
            {
	            var randomPlayer = Random.nextInt(neighbors.length);
	            actors(neighbors(randomPlayer)) ! pushSum(s,w)     
	        }
            
          }
          }
   }
}

object Project2 extends App {
     
   if(args.length != 3)
     println("Invalid Arguments")
     
     else if(args.length == 3)
     {
     var numNodes:Int = args(0).toInt
     var topology:String = args(1)
     var algorithm:String = args(2)
     var i:Int = 0
     var actors:List[ActorRef] = Nil
     
     var b:Long = 0
     val system = ActorSystem("GossipSimulator")
     val maxNoOfRumours = 10
     if(topology == "2d" || topology == "imp2d")
      while(math.sqrt(numNodes.toDouble)%1!=0){
       numNodes += 1
     }
     
     val numNodesRoot = math.sqrt(numNodes.toDouble)
     var superVisor:ActorRef = system.actorOf(Props[supervisor])
     
     while(i<numNodes){
        actors ::=  system.actorOf(Props[Nodes])
        i += 1
      }
   
   if(topology == "full" || topology == "Full")
   {
     var i:Int = 0
          while(i<actors.length){
            var neighbors:List[Int] = Nil
            
            var j:Int = 0
            while(j<actors.length){
              if(j!=i){
                neighbors ::= j
               		}
              j += 1
            }
            actors(i) ! Identity(i, actors, neighbors, maxNoOfRumours, superVisor)
            i += 1
          }
    }
   
   if(topology == "2D" || topology == "2d")
   {
    
     var i:Int = 0
          while(i<actors.length){
            var neighbors:List[Int] = Nil
            
            if((i.toDouble - numNodesRoot) >= 0)
              neighbors ::= ((i-numNodesRoot).toInt)
             if((i.toDouble + numNodesRoot) < actors.length)
             neighbors ::= ((i+numNodesRoot).toInt)
             if((i.toDouble+1) % numNodesRoot != 0)
               neighbors ::= i+1
             if((i.toDouble) % numNodesRoot != 0)
               neighbors ::= i-1  
            
            actors(i) ! Identity(i, actors, neighbors, maxNoOfRumours, superVisor)
            
            i += 1
          }
     }
   
   if(topology == "imp2D" || topology == "imp2d" || topology == "Imp2D" || topology == "Imp2d")
   {
     
     var i:Int = 0
     var j:Int = 0
     var k:Int = 0
     var randomNeighbor:Int = -1
          while(i<actors.length){
            var neighbors:List[Int] = Nil
            
            if((i.toDouble - numNodesRoot) >= 0)
              neighbors ::= ((i-numNodesRoot).toInt)
             if((i.toDouble + numNodesRoot) < actors.length)
             neighbors ::= ((i+numNodesRoot).toInt)
             if((i.toDouble+1) % numNodesRoot != 0)
               neighbors ::= i+1
             if((i.toDouble) % numNodesRoot != 0)
               neighbors ::= i-1  
            
            
            do{
              randomNeighbor = Random.nextInt(actors.length)
              for(k<-neighbors){
                if(randomNeighbor == k) randomNeighbor = -1
              }   
            }while(randomNeighbor == -1)
            
            neighbors ::= (randomNeighbor)
            
            actors(i) ! Identity(i, actors, neighbors, maxNoOfRumours, superVisor)
            
            i += 1
          }
     }
   
   if(topology == "line" || topology == "Line")
   {
          var i:Int = 0;
          while(i<actors.length){
            var neighbors:List[Int] = Nil
            
            if(i>0) neighbors ::= (i-1)
            if(i<actors.length-1) neighbors ::= (i+1)
            
            actors(i) ! Identity(i, actors, neighbors, maxNoOfRumours, superVisor)
            
            i += 1
          }
   }
    if(algorithm == "Gossip" || algorithm == "gossip"){
      b = System.currentTimeMillis
    		  superVisor ! KeepTime(b)
              actors(0) ! Gossip(-1)
     }
     if(algorithm == "push-sum" || algorithm == "Push-sum" || algorithm =="Pushsum" || algorithm == "pushsum"){

      b = System.currentTimeMillis
      superVisor ! KeepTime(b)
      
      var randomActor = Random.nextInt(actors.length);
	  actors(randomActor) ! pushSum(0,1)
	  
    }
    }
}


case class supervisor() extends Actor{
  
  var count:Double = 0
  var b:Long = 0
  var f:Int = 0
  var sumCheck:Int = 0
  
  def receive = {
    
    case KeepTime(b:Long) => {
      
      this.b = b
    }
    
    case Done(finished:Int,numNodes:Double) => {
      count += 1
      
      if(count == numNodes){
      //println(" All nodes Terminated ")
      println(" Time Taken = "+(System.currentTimeMillis()-b))
      sys.exit()
     }
    }
    
   case PushSumDone(actorId:Int,numNodes:Double,value:Int) => {
     this.f = value
      if(f==0)
      {
        count+=1    
      }
      if(f==1)
      {
        sumCheck = f
      }
      //if(count/numNodes && sumCheck == 1)             uncomment this if convergence not reached.
      if(count==numNodes && sumCheck==1){
      //println("All nodes heard rumor and s/w changed more than 3 consecutive rounds")  
      println(" Time Taken = "+(System.currentTimeMillis()-b))
      sys.exit();
      }
   }
  }
}
